### PR TITLE
DAOS-6352 control: Enforce uniform fault domain depth

### PR DESCRIPTION
**Pull request by kjacque**
- Reject attempts to join with a fault domain with a
  different number of levels than existing members in the
  membership DB.

Signed-off-by: Kris Jacque <kristin.jacque@intel.com>

### CODE REVIEW

**Review by mjmac — COMMENTED** [review]
Looks good to me. What does it look like when a server tries to join with a misconfigured fault domain? I'm guessing it tries to rejoin in a loop. We might want to look at that (perhaps as a followup) to instead exit with an error. It will still be restarted by systemd, at least a few times, but it might make the problem easier to spot in logs. Also, if the error returned is a fault.Fault, then the resolution message will be printed on daos_server's stderr.

**Comment by kjacque** [comment]
> Looks good to me. What does it look like when a server tries to join with a misconfigured fault domain? I'm guessing it tries to rejoin in a loop. We might want to look at that (perhaps as a followup) to instead exit with an error. It will still be restarted by systemd, at least a few times, but it might make the problem easier to spot in logs. Also, if the error returned is a fault.Fault, then the resolution message will be printed on daos_server's stderr.

The integrated behavior isn't really possible to test at the moment, since the control plane currently enforces a single-level fault domain, and there remains some work to be done to enable multi-level. Looking at the code, though, it looks like the engines would exit with errors, while the control plane would remain standing. Would we want the control plane to exit? I suppose that makes some sense since the admin will need to restart it to reconfigure the fault domain anyway, although in a cluster, maybe they'd want to be able to query the server to troubleshoot (via dmg).

**Comment by tanabarr** [comment]
> > Looks good to me. What does it look like when a server tries to join with a misconfigured fault domain? I'm guessing it tries to rejoin in a loop. We might want to look at that (perhaps as a followup) to instead exit with an error. It will still be restarted by systemd, at least a few times, but it might make the problem easier to spot in logs. Also, if the error returned is a fault.Fault, then the resolution message will be printed on daos_server's stderr.
> 
> The integrated behavior isn't really possible to test at the moment, since the control plane currently enforces a single-level fault domain, and there remains some work to be done to enable multi-level. Looking at the code, though, it looks like the engines would exit with errors, while the control plane would remain standing. Would we want the control plane to exit? I suppose that makes some sense since the admin will need to restart it to reconfigure the fault domain anyway, although in a cluster, maybe they'd want to be able to query the server to troubleshoot (via dmg).

If we specifically wanted administrator intervention then we might want to raise RAS event to get printed on syslog @ priority NOTICE? That's what I'm planning for example for others situations where administrator intervention is required e.g. waiting for SCM format.

**Review by mjmac — APPROVED** [review]

**Comment by kjacque** [comment]
> If we specifically wanted administrator intervention then we might want to raise RAS event to get printed on syslog @ priority NOTICE? That's what I'm planning for example for others situations where administrator intervention is required e.g. waiting for SCM format.

That's a good idea. The admin really needs to intervene to resolve the issue. I guess the MS would be the one generating the event? Since the affected system is unable to join.

**Comment by mjmac** [comment]
> 
> 
> > If we specifically wanted administrator intervention then we might want to raise RAS event to get printed on syslog @ priority NOTICE? That's what I'm planning for example for others situations where administrator intervention is required e.g. waiting for SCM format.
> 
> That's a good idea. The admin really needs to intervene to resolve the issue. I guess the MS would be the one generating the event? Since the affected system is unable to join.

Let's address that in a follow-on. We're going to discuss RAS at the upcoming WG to make sure we have a clear set of criteria for determining when something is RAS-worthy vs just a line in a log.

File: src/control/system/membership.go [hunk 1]
```diff
@@ -150,6 +160,10 @@ func (m *Membership) Join(req *JoinRequest) (resp *JoinResponse, err error) {
 		return nil, err
 	}
 
+	if err := m.checkReqFaultDomain(req); err != nil {
```
Comment (tanabarr) [inline src/control/system/membership.go:163]: don't we need to do the same check as above for count? also why don't we just do the check once before checking err == nil? I'm probably missing something obvious.

**Review by tanabarr — APPROVED** [review]

File: src/control/system/membership.go (hunk 1 above)
Comment (kjacque) [reply src/control/system/membership.go:163]: I started off that way, but then I found a special case where a member rejoins, but it's the only member. In that case the depth of the fault domain doesn't matter -- there are no other branches to compare it to, so it should be allowed to override the old fault domain.

**Pull request merged**